lib/stats/stats-query.c

```diff
@@ -117,62 +54,74 @@ _is_single_match(const gchar *key_str)
   return !strpbrk(key_str, "*?");
 }
 
-static GList *
-_query_counter_hash(const gchar *key_str)
+static void
+_process_counter_if_matching(StatsCluster *sc, gint type, StatsCounterItem *counter,
+                             gpointer user_data /*, gboolean* cancelled*/)
 {
-  GPatternSpec *pattern = g_pattern_spec_new(key_str);
-  GList *counters = NULL;
-  gpointer key, value;
-  GHashTableIter iter;
-  gboolean single_match;
-
-  _update_index();
-  single_match = _is_single_match(key_str);
-
-  g_mutex_lock(&stats_query_mutex);
-  g_hash_table_iter_init(&iter, counter_index);
-  while (g_hash_table_iter_next(&iter, &key, &value))
+  gpointer *args = (gpointer *) user_data;
+  int arg_ndx = 0;
+  GPatternSpec *pattern = (GPatternSpec *) args[arg_ndx++];
+  gboolean single_match = *(gboolean const *) args[arg_ndx++];
+  ProcessCounterCb process_func = (ProcessCounterCb) args[arg_ndx++];
+  gpointer process_func_user_data = args[arg_ndx++];
+  StatsFormatCb format_cb = (StatsFormatCb) args[arg_ndx++];
+  gpointer format_cb_user_data = args[arg_ndx++];
+  gboolean must_reset = *(gboolean const *) args[arg_ndx++];
+  gboolean *found = (gboolean *) args[arg_ndx++];
+
+  if (stats_cluster_is_alive(sc, type))
     {
-      if (_is_pattern_matches_key(pattern, key))
+      if (_is_pattern_matches_key(pattern, counter->name))
         {
-          StatsCounterItem *counter = (StatsCounterItem *) value;
-          counters = g_list_prepend(counters, counter);
+          *found = TRUE;
+
+          process_func(counter, process_func_user_data, format_cb, format_cb_user_data);
+          if (must_reset)
+            stats_control_reset_counter_if_needed(sc, counter);
 
           if (single_match)
-            break;
+            {
+              /* DISCUSS: cancelled parameter is not implemented yet in the counter enumerators, only in the cluster ones, and
+                          - not stopping here now does not make a performace difference as the original impl always interated over all the counter items in all the clusters
+                          - stopping here now will not update the indexed state (indexed_mask) of all of the clusters leaving them in an incosistent state (that might be acceptable
+                            as currently only this query module uses that state)
```

Comment:
This comment is no longer relevant, there is no indexed mask.

Reply:
removed the index mask related part